### DISCUSSION TITLE
Charm Segmentation Issues in Simnibs

### DISCUSSION THREAD

**Discussion by GiacomoBert** [simnibs/simnibs#131] [7.Creating head models] [answered]
Hi! Just wanted to double-click on this issue with charm not working well in slightly atrophic brains. 
It was posted on the simnibs-discussion mailinglist which is now deprecated.  I am posting down here the whole conversation, hopefully it will be useful for future users.

I am working with MRI scans of elderly and Alzheimer's disease patients and charm behaves the same way Joel  described (see discussion below), no matter how I tweak the parameters in the setting.ini. Citing joel , "the problem is that the cortical surface is smooth and pushed up right next to the csf boundary layer " 
Here an example from my data (fig 1) 
![fig_1](https://user-images.githubusercontent.com/39560621/232443414-c4262a56-60c9-429c-be2b-149dfcb34811.png)

Old headreco pipeline works very well though. 

I will add that I have tried to add a T2w "flair" image (since  it was the only T2w image I had) and the segmentation worked better for (most) of the healthy brain, while with (even slightly) atrophic brains it produced a fragmented GM layer (fig 2).
![fig_2](https://user-images.githubusercontent.com/39560621/232443513-9bba7230-245b-40e5-b5a2-ff3dddf0925b.png)

 It improves if I increase by 10-20 fold the parameter csf_factor (I guess because the t2w flair has the CSF signal suppress (fig 3), but the GM layer still has holes.

![fig_3](https://user-images.githubusercontent.com/39560621/232443626-76986717-2cd2-40fd-ac69-4d85c0ea0e4d.png)

Do you have any additional guidelines for making charm work better with atrophic brains? 

here the figures in gdrive: 
https://drive.google.com/drive/folders/1EfVXcMSi19I2iyql8jkX6wosEIOXsQ9U?usp=sharing

Thanks for your help! 

Giacomo

------------------------------------------------------------------------------------------------------

```
Il giorno ven 12 ago 2022 alle ore 09:54 <[simnibs-discuss-request@drcmr.dk](mailto:simnibs-discuss-request@drcmr.dk)> ha scritto:
Send Simnibs-discuss mailing list submissions to
        [simnibs-discuss@drcmr.dk](mailto:simnibs-discuss@drcmr.dk)

To subscribe or unsubscribe via the World Wide Web, visit
        https://mailman.drcmr.dk/mailman/listinfo/simnibs-discuss
or, via email, send a message with subject or body 'help' to
        [simnibs-discuss-request@drcmr.dk](mailto:simnibs-discuss-request@drcmr.dk)

You can reach the person managing the list at
        [simnibs-discuss-owner@drcmr.dk](mailto:simnibs-discuss-owner@drcmr.dk)

When replying, please edit your Subject line so it is more specific
than "Re: Contents of Simnibs-discuss digest..."


Today's Topics:

   1. Re: Charm Segmentation Issues in Simnibs 4 Beta (Oula)


----------------------------------------------------------------------

Message: 1
Date: Thu, 11 Aug 2022 08:19:46 +0200
From: Oula <[oulap@drcmr.dk](mailto:oulap@drcmr.dk)>
To: Joel Upston <[jupston@unm.edu](mailto:jupston@unm.edu)>
Cc: "[discuss@simnibs.org](mailto:discuss@simnibs.org)" <[discuss@simnibs.org](mailto:discuss@simnibs.org)>, "[support@simnibs.org](mailto:support@simnibs.org)"
        <[support@simnibs.org](mailto:support@simnibs.org)>
Subject: Re: [Simnibs-discuss] Charm Segmentation Issues in Simnibs 4
        Beta
Message-ID: <[b7414acc-7776-caa8-62fa-6172f4c9fb2d@drcmr.dk](mailto:b7414acc-7776-caa8-62fa-6172f4c9fb2d@drcmr.dk)>
Content-Type: text/plain; charset="utf-8"; Format="flowed"

Hi Joel,

I'd be interested to look at cases 1) and 2). I didn't have any subjects
with severe atrophy to test on so I think these can certainly be
improved. Also having the cortical GM pushed all the way up to the CSF
surface sounds strange.

W.r.t. fixing the affine: there are examples on how to do this in the
documentation. If you click through tutorial -> Creating Head Models and
then go all the way down on that page and click on "Fixing the Affine
Registration". This should give you some pointers on how to do that.
I'll also implement a flag for passing a registration matrix and
skipping the optimization.

Best,

Oula


On 8/10/22 17:56, Joel Upston wrote:
> Oula,
>
> Thank you for the explanation and I found as you said that ?mni? is
> faster but does seem to fail more than ?atlas?. However ?after more
> testing with these settings it looks like it is still having issues
> about 40% of the time, this is over about 20 subjects I have tested
> some with T1 and T2 others with just T1. Some of the issues are
> summarized as follows:
>
> 1) The template appears to be registered correctly and the tissue
> types are correctly segmented except the gray matter which as I
> mentioned before is ?pushed? right to the edge of the csf inner layer.
> In these cases running the same scan through headreco with cat12 seems
> to generate the correct gm surface but then have to use outside
> programs to combine the two segmentations via editing, use the
> add_tissues_to_upsampling and then rerun the meshing.
>
> 2) Some of the population of this dataset is above 50, with some
> atrophy and large ventricles. The ventricles that are often large,
> fail in the charm segmentation. My solution is to again use the
> headreco segmentation ventricles in the charm segmentation.
>
> 3) The probabilistic atlas is still not in registration with the T1,
> or could be improved. Would it be possible to add in a 12 parameter
> registration matrix (similar to the init transform) that it is forced
> to used, that way if another outside program such as AFNI or spm gets
> a better registration I can use that? I have noticed that using the
> init transform with the proper alignment ends up finding the same
> incorrect minimum in the registration optimization.
>
> Overall I am happy to share some of this data as well to help with
> some of these issues. Thank you for all your help and all your
> contributions to automate a more accurate head model for the field.
>
>
> Thank you,
>
> Joel Upston
>
>> On Aug 8, 2022, at 12:30 AM, Oula <[oulap@drcmr.dk](mailto:oulap@drcmr.dk)> wrote:
>>
>> ?
>>
>> *? [EXTERNAL]*
>>
>> Hi Joel,
>>
>> The standard init_type ("mni") does a registration between the T1w
>> MNI template and the input scan using a correlation metric. This
>> assumes that the tissue intensities of the input scan and the MNI
>> template are more or less the same, and it can sometimes fail if
>> that's not the case (even if the input is a T1w scan).
>>
>> Init_type="atlas" does a registration between the input and a
>> probabilistic atlas using a mutual information-based metric (similar
>> to what SPM12 does). This metric doesn't care about the specific
>> tissue intensities of the input, so you could, for example, feed in a
>> T2w scan and it should still work.
>>
>> The reason I set "mni" as default is that it's faster and seemed to
>> work well on the cases I tested it. I'll try to do more testing on
>> different T1w scans to see if we should set the "atlas"-option as
>> default.
>>
>> Best,
>>
>> Oula
>>
>>
>> On 8/5/22 18:12, Joel Upston wrote:
>>> ? Oula,
>>>
>>> Thank you for the quick response. Yes the template alignment seems
>>> to be a bit better from the affine from headreco, than that of what
>>> is given through the charm optimization. I tried the noneck which
>>> didn?t change it but just tried the ?atlas? init_type parameter and
>>> that worked. How exactly is it different?
>>>
>>> Thank you,
>>>
>>> Joel Upston
>>>
>>>> On Aug 5, 2022, at 5:18 AM, Oula <[oulap@drcmr.dk](mailto:oulap@drcmr.dk)> wrote:
>>>>
>>>> ?
>>>>
>>>> *[EXTERNAL]*
>>>>
>>>> Hi Joel,
>>>>
>>>> Did you check the template alignment after forcing the affine from
>>>> headreco? You can also try switching the init_type from "mni" to
>>>> "atlas" in the [samseg] section of the settings file. It also looks
>>>> like the scans don't have neck so please include the --noneck flag
>>>> if that's the case.
>>>>
>>>> If you've tried all these things already, I'm happy to have a look
>>>> if you're allowed to share a scan or two from the data set.
>>>>
>>>> Best,
>>>> Oula
>>>>
>>>>
>>>> On 8/5/22 12:32, Joel Upston wrote:
>>>>> ?
>>>>> Hi there,
>>>>>
>>>>> I have worked with the Simnibs 3.2.6 headreco to do segmentation
>>>>> and now trying out the charm segmentation routine in the Simnibs 4
>>>>> Beta release. I have multiple subjects (about 10 subjects tried so
>>>>> far) that have looked good in the headreco segmentation in terms
>>>>> of the brain segmentation especially using the T1 and T2 but when
>>>>> I try this with the charm segmentation, I run into a problem about
>>>>> 50% of the time. The problem is that the cortical surface is
>>>>> smooth and pushed up right next to the csf boundary layer, but the
>>>>> wm surface looks good. I have attached an image of what is
>>>>> happening to show a better description. I have made sure that the
>>>>> template is aligned by forcing in the code to use the Simnibs 3
>>>>> warp from MNI2Conform 12 parameter alignment, where the previous
>>>>> headmodel seemed to be ?accurate. I also try to change some of the
>>>>> settings in the settings.ini file such as all parameters in the
>>>>> segment section and th_initial in the surface section with no
>>>>> avail. As I run the same subject in the Simnibs 3.2.6 the
>>>>> headmodel looks good, I have tried to edit via switching out the
>>>>> brain+csf segmentation from simnibs3 to simnibs 4 and reran the
>>>>> mesh after editing. This works reasonably well but the some of the
>>>>> external to brain structures are affected, plus requires double
>>>>> the runtime to run versions 3 and 4. Have you noticed a problem
>>>>> similar to this and do you have any recommendations to try? Thank
>>>>> you for all your work and I appreciate the research to keep
>>>>> improving this modeling.
>>>>>
>>>>> Thank you,
>>>>>
>>>>> Joel Upston
>>>>>
>>>>> Correct_GM_surface_from_headreco.png
>>>>>
>>>>> Incorrect_GM_surface_from_charm.png
>>>>> Screenshot%20from%202022-08-04%2012-51-01.png
-------------- next part --------------
An HTML attachment was scrubbed...
URL: <https://mailman.drcmr.dk/pipermail/simnibs-discuss/attachments/20220811/74195ad1/attachment.html>
-------------- next part --------------
A non-text attachment was scrubbed...
Name: Correct_GM_surface_from_headreco.png
Type: image/png
Size: 128092 bytes
Desc: not available
URL: <https://mailman.drcmr.dk/pipermail/simnibs-discuss/attachments/20220811/74195ad1/attachment.png>
-------------- next part --------------
A non-text attachment was scrubbed...
Name: Incorrect_GM_surface_from_charm.png
Type: image/png
Size: 104518 bytes
Desc: not available
URL: <https://mailman.drcmr.dk/pipermail/simnibs-discuss/attachments/20220811/74195ad1/attachment-0001.png>
-------------- next part --------------
A non-text attachment was scrubbed...
Name: Screenshot%20from%202022-08-04%2012-51-01.png
Type: image/png
Size: 370088 bytes
Desc: not available
URL: <https://mailman.drcmr.dk/pipermail/simnibs-discuss/attachments/20220811/74195ad1/attachment-0002.png>

------------------------------

Subject: Digest Footer

_______________________________________________
Simnibs-discuss mailing list
[Simnibs-discuss@drcmr.dk](mailto:Simnibs-discuss@drcmr.dk)
https://mailman.drcmr.dk/mailman/listinfo/simnibs-discuss


------------------------------

End of Simnibs-discuss Digest, Vol 33, Issue 11
```

**Comment by oulap** [comment, marked as answer]
Hi,

We haven't tested the pipeline specifically on atrophied brains, but it would make sense that boosting the csf factor helps. If I remember Joel's case correctly, he's problem was a combination of the affine initialization being a bit off (you can check this in the charm_report.html) and the segmentation parameters not being exactly optimal for the type of scans he had.

I provided him with a slightly tweaked settings file along with a new parameter file for T1+T2  segmentation (see attached). You can pass this settings file to charm by using the --usesettings flag. The gmm parameter file should be in the same folder as the custom settings file, if it is not charm will default to the standard parameter file shipped with simnibs. Also if you have some example scans you can share, I can try to change the settings so that the 
GM behaves better.

[settings+parameter_files.zip](https://github.com/simnibs/simnibs/files/11250277/settings%2Bparameter_files.zip)

**Comment by GiacomoBert** [comment]
@oulap , thanks for your fast reply. With those parameters the extractions works way better in both elderly brains and Alzheimer's disease brains! 

> Also if you have some example scans you can share, I can try to change the settings so that the
> GM behaves better.

Unfortunatelly I cannot share this data yet for privacy issues, but I am working to solve them. I will notify you when they are online. 


Thanks again!

**Comment by carysbennett** [comment]
I have been running T1 images through charm and it always crashes at the same point so I am unsure what to do next. I have included the charm report below. 
![image](https://user-images.githubusercontent.com/132445339/236333685-17f0adf2-d43c-451a-a891-8e2083edaae7.png)
![image](https://user-images.githubusercontent.com/132445339/236334604-6e90bb75-5519-4dee-8e73-d4d1bcd1faac.png)

**Reply by oulap** [reply]
Can you run only the first step (atlas registration) like this:

charm subid T1.nii.gz --initatlas

and then inspect the output of charm_report.html (note not charm_log.html). charm_report has a viewer, which allows you to check if the initial template registration was successful. I'm guessing that might have failed and is causing problems later in the segmentation pipeline. Below is an example where the registration looks good. We have some examples of failure cases here: 

https://simnibs.github.io/simnibs/build/html/tutorial/advanced/fix_affine_registration.html#fix-affine-registration-tutorial

![image](https://user-images.githubusercontent.com/32451236/236338979-41ef6697-d56f-4565-be4d-fc69ba77a973.png)

**Comment by carysbennett** [comment]
Hi,
    I ran the charm command on the atlas and it is still having problems. It is saying invalid axises. Then, when I try and look at the atlas charm says its using it won't let me cd into it or read the output file. Also only the charm log, not charm report is showing up on it. I have included screenshots below.
Thank you so much for all your help,
Carys
 [cid:3dab9436-a0c4-4616-a3d4-f4c14a3c1350]

________________________________

**Comment by oulap** [comment]
It seems the screenshots didn't come through. You can also send those to support@simnibs.org.
Sounds like there might be something going on with the input scan. Are you allowed to share this example? It would make debugging easier. If yes, you can send a download link to the support email.  If not, could you share the header information of the scan? You can get that for example using the FSL command fslhd.

**Comment by oulap** [comment]
We have a new pre-release available here:

https://github.com/simnibs/simnibs/releases/tag/dev

In this version you can use surfaces from an existing freesurfer run to improve the gray matter in the charm head model. At the moment you would need to run freesurfer separately on the same subject and point charm to the freesurfer output folder. In the future, we'll integrate improved surface models directly into charm so that you don't need to run FS separately.
